server/src/main/java/org/elasticsearch/common/component/AbstractLifecycleComponent.java

```diff
@@ -52,59 +49,64 @@ public void removeLifecycleListener(LifecycleListener listener) {
 
     @Override
     public void start() {
-        if (!lifecycle.canMoveToStarted()) {
-            return;
-        }
-        for (LifecycleListener listener : listeners) {
-            listener.beforeStart();
-        }
-        doStart();
-        lifecycle.moveToStarted();
-        for (LifecycleListener listener : listeners) {
-            listener.afterStart();
+        synchronized (lifecycle) {
```

Comment:
I think it can become tricky to synchronize on an object that can also be directly accessed by subclasses. The state of `Lifecycle` has only one field that is declared `volatile` and as far as I could see all usages in subclasses only query the state and only the base class modifies it. From that perspective we are safe but it is easy to introduce subtle bugs. I wonder whether in the future it would make sense to think about encapsulating `Lifecycle` in `AbstractLifecycleComponent` to make this a bit more robust.